### ISSUE TITLE
Record expressions do not respect renaming of fields

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#4720]
I'm getting some very weird behaviour related to field names in record expressions:
```agda
open import Data.Bool
open import Data.Nat
open import Data.Product
open import Relation.Binary.PropositionalEquality

works : Σ Bool λ _ → ℕ
works = record{ fst = true; snd = 0 }

fails : Σ Bool λ _ → ℕ
fails = record{ proj₁ = true; proj₂ = 0 }
```
The first definition works, despite `fst` and `snd` not being in scope. The second definition raises a warning:
```
The record type Σ does not have the fields proj₁, proj₂ but it
would have the fields fst, snd
when checking that the expression
record { proj₁ = true ; proj₂ = 0 } has type Σ Bool (λ _ → ℕ)
```

**Comment by andreasabel** [comment]
Record labels bypass the scope.  They are simply concrete names.
https://github.com/agda/agda/blob/d1cc9f32e7e5e4ee25cc7e66dd7304ed31659087/src/full/Agda/Syntax/Abstract.hs#L28
https://github.com/agda/agda/blob/d1cc9f32e7e5e4ee25cc7e66dd7304ed31659087/src/full/Agda/Syntax/Abstract.hs#L104-L105
https://github.com/agda/agda/blob/d1cc9f32e7e5e4ee25cc7e66dd7304ed31659087/src/full/Agda/Syntax/Abstract.hs#L125-L129
I guess that when record(expression)s were retrofitted on the original design of Agda, the puzzle of how to manage record field names was not solved: They should somehow not needed to be explicitly imported, since they are determined by the record type, but on the other hand, just using the original field names stored with the record type does not fly (as you observed in the OP --- however, it did not bother us much before we had `Agda.Builtin.Sigma`).  Never wondered why concrete names for the fields are stored in the `Record` structure in the signature?  They should absolutely have no place there!

One way to get hold of the up-to-date record field names is to ask the record module associated with the record type.  However, note that the record module need not be in scope either.  Some loophole needs to be implemented here.

If you are worried about the OP, you should be abhorred by the code for record expression _inference_ (slightly sanitized by #4267): 
https://github.com/agda/agda/blob/d1cc9f32e7e5e4ee25cc7e66dd7304ed31659087/src/full/Agda/TypeChecking/Records.hs#L198-L214
[This happens when we let AIM participants contribute their code sprint "feature". 🤣]
I ranted about the record expression inference already in #4267 (and more generally in #4272).

**Comment by andreasabel** [comment]
MWE:
```agda
record R : Set₂ where
  field
    old : Set₁

open R public renaming (old to new)

test : R
test = record { new = Set }
```

**Comment by jespercockx** [comment]
> If you are worried about the OP, you should be abhorred by the code for record expression inference (slightly sanitized by #4267):

You may consider me suitably horrified.

**Comment by andreasabel** [comment]
For my MWE, what is going on in the scope?
1. We have the record module `R`:
```agda
module R (self : R) where
  old : Set1
  old = self .old
```
2. We open it, this gives a new projection name in scope
```agda
new : R -> Set1
new = R.old
```
3. Then we build `record { new = Set } : R`.

The question is how we should handle the record labels, here `old` and `new`.

If we want this renaming of labels feature, we could not continue our feature that the record labels need not be in scope to construct a record with it.  Otherwise, how would we deal with this:
```agda
record R (A : Set) : Set where
  field a b : A

open R renaming (a to b; b to a)

test : R Bool
test = record { a = true; b = false }
```
Would the original labels be used here or the new ones?

So if we abandon the "labels need not be in scope" feature, we could treat a record expression with the same scoping rules as the definition-by-copattern.
```agda
test .a = true
test .b = false
```

The flaw in the ointment is that we currently do not have overloaded projections with unopened record modules.  An it is not so easy to get this because of the conflict with the use or field names as field values if you `open R r`.

Maybe a different idea is to have a `renaming` directives that renames a record type _with_ its field labels.
```agda
open M renaming (record R to S; R.a to S.b; R.b to S.a)
```
Then, if we construct `record {a = true; b = false}` we have to find a record type in scope (`S`) that has fields `a` and `b`.

**Comment by JacquesCarette** [comment]
There are reasons why I'm very interested in type theories with first-class names (for record and modules) and renaming...

Note that CASL, Hets, Maude and other specification language in the OBJ family seem to deal with such things reasonably well (and I think SpecWare and DesignWare did too). Isabelle has some features, but I don't know how robust they are.

My suspicion is that this is more than just an implementation/design issue, it needs a bit of theory to make it all sensible. Why? Because in the above you have not yet dealt with **composition** of renames. And that throw another new level of complications into the mix - and such composition is already available as a feature in Agda (i.e. do some `open public` with renames in one module, and then rename again in a later module). You also have 'diamond' problems lurking as well.

**Comment by andreasabel** [comment]
> You also have 'diamond' problems lurking as well.

Yeah, and we do not have similar problems with constructors or other names.  E.g. for constructors, it does not matter how many aliases we have for the same original constructor.  However, for record expressions it would look very awkward if one took some field names from one renaming and other from another renaming (in particular, without having to have the names imported).
The drastic solution is to require the projections to be in scope to form a record expression, just like for copattern matches.

Unfortunately, I do not know the systems you mentioned and how they deal with this.

**Comment by JacquesCarette** [comment]
I do have quite a bit more that I could say about this [I do know the details of how some of these other systems handle this; also, I've written a paper about this exact problem], but doing so in a comment seems sub-optimal. I'd be glad to have some online chat.
